feat: Add new storage account and related new resources.

VaishnaviLende opened the pull request:
## Description

<!--Why this PR? What is changed? What is the effect? etc.-->

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

<!--Run through the checklist if your PR adds a new module.-->

- [ ] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

<!--Run through the checklist if your PR updates an existing module.-->

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.
- [x] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [x] I have updated the examples in README with the latest module version number.

VaishnaviLende commented:
> 



> @VaishnaviLende please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Nuance"

alex-frankel commented on modules/storage/storage-account/README.md:
```diff
@@ -17,15 +17,46 @@ The output of the module is the ID of the created or existing Storage Account, w
+| `ruleId`                          | `string` | No       | Rule Id is auto-generated for each new rule on destination account. It is required for put policy on source account.                 |
+| `policyId`                        | `string` | No       | This is the name to provide for objectReplicationPolicies.                                                                           |
+| `objectReplicationPolicy`         | `bool`   | No       | When performing object replication, it must be true and all resources necessary for the destination storage account will be created. |
+| `managedIdentityName`             | `string` | Yes      | User defined name to provide userAssignedIdentities resource.                                                                        |
```
Can you expand on why managedIdentity params are required? Can these be optional?

VaishnaviLende replied:
Yes, this can be optional and added default value to this param and also updated in readme.md file.

alex-frankel commented on modules/storage/storage-account/README.md:
```diff
@@ -17,15 +17,46 @@ The output of the module is the ID of the created or existing Storage Account, w
+| `destBlobType`                    | `string` | No       | Specifies the type of blob to manage the lifecycle policy.                                                                           |
+| `destChangeFeedEnabled`           | `bool`   | No       | Indicates whether change feed event logging is enabled for the Blob service.                                                         |
+| `destVersioningEnabled`           | `bool`   | No       | Versioning is enabled if set to true. To the destination storage account, set true.                                                  |
+| `destSkuName`                     | `string` | Yes      | The SKU name required for account creation; optional for update.                                                                     |
```
Can you expand on why this param is required? As well as how it can be optional for an update? Bicep deployments should not distinguish between create and update

VaishnaviLende replied:
Assigned default value to this param and now it is optional.

alex-frankel reviewed (commented):
Initially would like to understand the context for the additional required parameters

alex-frankel commented:
@dciborow - can you take a look at this one? I don't have enough context on the Storage RP to review this one effectively

dciborow commented:
> @dciborow - can you take a look at this one? I don't have enough context on the Storage RP to review this one effectively

Overall, the improvements look good, but they need to be merged into the existing module a little bit more.

@VaishnaviLende , 
1. try to avoid adding any required parameters, by providing default values.
2. Merge 'destinationStorageAccount' into 'storageAccount'
3. Use the nested resource notation for the children resources

dciborow reviewed (dismissed):
1. try to avoid adding any required parameters, by providing default values. Dedup things like "destLocation", and just use the existing 'location' instead.
2. Merge 'destinationStorageAccount' into 'storageAccount'
3. Use the nested resource notation for the children resources

VaishnaviLende commented:
> 
Thanks for the review comments. I am working on point 1 and 3. 

On point 2 - destinationStorageAccount is created only if object replication policy is enabled. Are you suggesting to create it as nested resource under storage account?

VaishnaviLende commented:
> > 
> 
> Thanks for the review comments. I am working on point 1 and 3.
> 
> On point 2 - destinationStorageAccount is created only if object replication policy is enabled. Are you suggesting to create it as nested resource under storage account?

The changes for points 1 and 3 have been completed and committed. Could you kindly review it once and let me know if any changes are necessary. Thanks

dciborow commented:
> > 
> 
> Thanks for the review comments. I am working on point 1 and 3.
> 
> On point 2 - destinationStorageAccount is created only if object replication policy is enabled. Are you suggesting to create it as nested resource under storage account?

Let me take a fresh pass at this with this understanding.

dciborow commented on modules/storage/storage-account/main.bicep:
```diff
@@ -19,6 +19,100 @@ param isZoneRedundant bool = false
+param managedIdentityName string = 'MyManagedIdentity'
+
+@description('It specifies the name or Id of the Role Definitions.')
+param roleDefinitionIdOrName string
```
This param should include a default.

VaishnaviLende replied:
Added default value as Reader.

dciborow commented on modules/storage/storage-account/main.bicep:
```diff
@@ -19,6 +19,100 @@ param isZoneRedundant bool = false
+param destPublicNetworkAccess string = 'Disabled'
+
+@description('Set the minimum TLS version to be permitted on requests to storage. The default interpretation is TLS 1.0 for this property.')
+param destMinimumTlsVersion string = minimumTlsVersion
```
should be able to use an @Allowed here for the allowed versions.

VaishnaviLende replied:
@allowed added here.

dciborow commented on modules/storage/storage-account/main.bicep:
```diff
@@ -19,6 +19,100 @@ param isZoneRedundant bool = false
+param publicNetworkAccess string = 'Disabled'
+
+@description('Set the minimum TLS version to be permitted on requests to storage. The default interpretation is TLS 1.0 for this property.')
+param minimumTlsVersion string = 'TLS1_2'
```
should be able to use an @Allowed here for the allowed versions.

VaishnaviLende replied:
@allowed added here.

dciborow commented on modules/storage/storage-account/main.bicep:
```diff
@@ -19,6 +19,100 @@ param isZoneRedundant bool = false
+param allowCrossTenantReplication bool = false
+
+@description('Allow or disallow public network access to Storage Account. Value is optional but if passed in, must be Enabled or Disabled.')
+param publicNetworkAccess string = 'Disabled'
```
use @Allowed for Enabled/Disabled

VaishnaviLende replied:
@allowed added here.

dciborow commented on modules/storage/storage-account/README.md:
Could you add an example here using the destination storage account?

VaishnaviLende replied:
I have added a test case [please refer to test2] as an example of a destination storage account. 
I am attaching screenshots as reference.
1. The object replication policy is enabled, and two storage accounts, source and destination, are required. Two storage accounts, source and destination, were created in test 2. 
![image](https://github.com/Azure/bicep-registry-modules/assets/113136137/1ec6f1da-bb5c-4d5c-b14d-275e77c44e35)

2. As a result, as we can see below on the Azure portal 
![image](https://github.com/Azure/bicep-registry-modules/assets/113136137/1bcb1087-6ad8-445c-b681-f532d0aae131)

dciborow commented on modules/storage/storage-account/test/main.test.bicep:
```diff
@@ -13,10 +19,50 @@ param location string = resourceGroup().location
+}
+
+//Test 0.
 module test0 '../main.bicep' = {
```
create a new test case for this. We want to keep test0 as is requiring the minimum amount of parameters.

VaishnaviLende replied:
Added two more test cases, one for when objectReplicationPolicy is disabled and one for when it is enabled, keeping test0 intact. Please refer to test 1 and 2, respectively.

VaishnaviLende commented:
> 1. try to avoid adding any required parameters, by providing default values. Dedup things like "destLocation", and just use the existing 'location' instead.
> 2. Merge 'destinationStorageAccount' into 'storageAccount'
> 3. Use the nested resource notation for the children resources

I addressed all of the review comments. Could you kindly review the changes once and approve them?

Ankurmutha commented:
@dciborow - Hope you are doing well! Request you to approve this PR.

Ankurmutha commented:
> shenglol

@shenglol - Is it possible for you to review and approve this PR in absence of @dciborow ?

shenglol commented:
Could you resolve the conflicts for the PR?

VaishnaviLende commented:
> Could you resolve the conflicts for the PR?

@shenglol - Thank you for your response. I have resolved all the conflicts. Could you please review and merge this pull request?

shenglol commented:
@VaishnaviLende Looks like the REAME file needs a regeneration.

![image](https://github.com/Azure/bicep-registry-modules/assets/16367959/e55db89b-f918-4d37-9e7b-915e3094ef90)

VaishnaviLende commented:
> @VaishnaviLende Looks like the REAME file needs a regeneration.
> 
> ![image](https://user-images.githubusercontent.com/16367959/244796715-e55db89b-f918-4d37-9e7b-915e3094ef90.png)

Updated readme file.

VaishnaviLende commented:
> > @VaishnaviLende Looks like the REAME file needs a regeneration.
> > ![image](https://user-images.githubusercontent.com/16367959/244796715-e55db89b-f918-4d37-9e7b-915e3094ef90.png)
> 
> @shenglol - Updated readme file.

@shenglol - I updated bicep version and also readme, main.json.

shenglol commented:
> > > @VaishnaviLende Looks like the REAME file needs a regeneration.
> > > ![image](https://user-images.githubusercontent.com/16367959/244796715-e55db89b-f918-4d37-9e7b-915e3094ef90.png)
> > 
> > 
> > @shenglol - Updated readme file.
> 
> @shenglol - I updated bicep version and also readme, main.json.

Can you also update `brm` with `dotnet tool update --global Azure.Bicep.RegistryModuleTool`? I think there are some improvements that may affect README generation.

VaishnaviLende commented:
> > > > @VaishnaviLende Looks like the REAME file needs a regeneration.
> > > > ![image](https://user-images.githubusercontent.com/16367959/244796715-e55db89b-f918-4d37-9e7b-915e3094ef90.png)
> > > 
> > > 
> > > @shenglol - Updated readme file.
> > 
> > 
> > @shenglol - I updated bicep version and also readme, main.json.
> 
> Can you also update `brm` with `dotnet tool update --global Azure.Bicep.RegistryModuleTool`? I think there are some improvements that may affect README generation.

@shenglol - I updated brm also.
